src/data/AmendmentCenter.cpp

```diff
@@ -0,0 +1,160 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "data/AmendmentCenter.hpp"
+
+#include "data/BackendInterface.hpp"
+#include "data/Types.hpp"
+#include "util/Assert.hpp"
+
+#include <boost/asio/spawn.hpp>
+#include <fmt/compile.h>
+#include <ripple/basics/Slice.h>
+#include <ripple/basics/base_uint.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/SField.h>
+#include <ripple/protocol/STLedgerEntry.h>
+#include <ripple/protocol/Serializer.h>
+#include <ripple/protocol/digest.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <ranges>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace {
+std::vector<std::string_view> SUPPORTED_AMENDMENTS = {};
+}  // namespace
+
+namespace data {
+
+AmendmentKey::AmendmentKey(std::string_view amendmentName) : name(amendmentName)
+{
+    SUPPORTED_AMENDMENTS.push_back(amendmentName);
+}
+
+AmendmentKey::AmendmentKey(std::string const& amendmentName) : name(amendmentName)
+{
+    SUPPORTED_AMENDMENTS.push_back(amendmentName);
+}
+
+AmendmentKey::operator std::string() const
+{
+    return std::string{name};
+}
+
+AmendmentKey::operator ripple::uint256() const
+{
+    return Amendment::GetAmendmentId(name);
+}
+
+AmendmentCenter::AmendmentCenter(std::shared_ptr<data::BackendInterface> const& backend) : backend_{backend}
+{
+    namespace rg = std::ranges;
+    namespace vs = std::views;
+
+    rg::copy(
+        ripple::allAmendments() | vs::transform([&](auto const& p) {
+            auto const& [name, support] = p;
+            return Amendment{
+                .name = name,
+                .feature = Amendment::GetAmendmentId(name),
+                .isSupportedByXRPL = p.second != ripple::AmendmentSupport::Unsupported,
+                .isSupportedByClio = rg::find(SUPPORTED_AMENDMENTS, p.first) != rg::end(SUPPORTED_AMENDMENTS),
+                .isRetired = p.second == ripple::AmendmentSupport::Retired
+            };
+        }),
+        std::back_inserter(all_)
+    );
+
+    for (auto const& am : all_ | vs::filter([](auto const& am) { return am.isSupportedByClio; }))
+        supported_.insert_or_assign(am.name, am);
+}
+
+bool
+AmendmentCenter::isSupported(std::string name) const
+{
+    return supported_.contains(name) && supported_.at(name).isSupportedByClio;
+}
+
+std::map<std::string, Amendment> const&
+AmendmentCenter::getSupported() const
+{
+    return supported_;
+}
+
+std::vector<Amendment> const&
+AmendmentCenter::getAll() const
+{
+    return all_;
+}
+
+bool
+AmendmentCenter::isEnabled(std::string name, uint32_t seq) const
+{
+    return data::synchronous([this, name, seq](auto yield) { return isEnabled(yield, name, seq); });
```

Comment:
I might get it wrong, can not see the call path clearly from diff view.  isEnabled (line 115) will call the overload isEnabled (line 121), then AmendmentKey's constructor will be called implicitly, then insert something to the SUPPORTED_AMENDMENTS.

Reply:
Probably to make compiler find such issues it is better to add `explicit` to AmendmentKey's constructors.

Reply:
No, the purpose of this proxy object is to allow seamless conversions. I just forgot to split the one that writes to SUPPORTED_AMENDMENTS out to a separate subclass - my bad. Great find @cindyyan317